### PR TITLE
Add Barbuta

### PR DESCRIPTION
**Pull request by ScipioWright**
Adds the Barbuta folder and its files. Some of this will be dependent on what ends up in the actual world class, but there's no world class here yet so

### CODE REVIEW

File: worlds/ufo50/barbuta/items.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    classification: IC
+    quantity: int
+
+
+barbuta_item_table: Dict[str, ItemInfo] = {
+    "$50": ItemInfo(0, IC.progression, 5),
+    "$100": ItemInfo(1, IC.progression, 5),
+    "Umbrella": ItemInfo(2, IC.progression, 1),
+    "Necklace": ItemInfo(3, IC.progression, 1),
+    "Pin": ItemInfo(4, IC.progression | IC.useful, 1),
```
Comment (LeonarthCG) [inline worlds/ufo50/barbuta/items.py:24]: Just so we stay on the same page, why are some of the items progression, but others progression + useful?
As in, what is the difference that made you go with one or the other?

File: worlds/ufo50/barbuta/locations.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+barbuta_location_table: Dict[str, LocationInfo] = {
+    "Green Skull - A1": LocationInfo(0, "Platforms above D4"),  # $100
+    "Egg Shop - B6": LocationInfo(1, "Boss Area"),  # $100 each
+    "Upper Shop Candy - C1": LocationInfo(2, "Platforms above D4"),  # costs $100
+    "Upper Shop Umbrella - C1": LocationInfo(3, "Platforms above D4"),  # idk the price, probably $100
+    "Chest - C2": LocationInfo(4, "Platforms above D4"),  # Pin chest
+    "Bat Altar - D1": LocationInfo(5, "Bat Altar"),  # bat noises
+    "Coin - D4": LocationInfo(6, "Starting Area"),  # $50, breakable block
+    "Little Guy Breaks a Wall - D7": LocationInfo(7, "C7 above Ladders"),  # costs $500
```
Comment (LeonarthCG) [inline worlds/ufo50/barbuta/locations.py:30]: I'll be honest, I'm not a fan of "Broken Wall"/"Breaks a Wall" but I'm not sure what else we would call it.
Hammer Guy? Backdoor Open? Dunno.

File: worlds/ufo50/barbuta/items.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    "Candy": ItemInfo(5, IC.progression, 1),
+    "Wand": ItemInfo(6, IC.progression | IC.useful, 1),
+    "Blood Sword": ItemInfo(7, IC.progression, 1),
+    "Key": ItemInfo(8, IC.progression, 1),
+    "Bat Soul": ItemInfo(9, IC.progression, 1),
```
Comment (LeonarthCG) [inline worlds/ufo50/barbuta/items.py]: The internal name for this seems to be "Bat Orb" but yours sounds cooler lol

File: worlds/ufo50/barbuta/rules.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    set_rule(world.get_location("Barbuta - Lower Shop Umbrella - F2"),
+             rule=lambda state: has_money(100, state, player))
+    set_rule(world.get_location("Barbuta - Lower Shop Trash - F2"),
+             rule=lambda state: has_money(50, state, player))
+    set_rule(world.get_location("Barbuta - Lower Shop Pin - F2"),
+             rule=lambda state: has_money(200, state, player))
+    set_rule(world.get_location("Barbuta - Little Guy Breaks a Wall"),
+             rule=lambda state: has_money(500, state, player))
+    set_rule(world.get_location("Barbuta - Bat Altar"),  # todo: should we do the wand version too?
+             rule=lambda state: state.has(blood_sword, player))
```
Comment (LeonarthCG) [inline worlds/ufo50/barbuta/rules.py]: I don't think the blood sword is needed for the altar at all, I just went to double check and got it to spawn with both the regular sword and the wand.

I can confirm that having the wand makes the orb spawn in the blood sword chest though, and that this is the only special case in the chest code (technically it's in the item drop's creation code I guess? whatever).

File: worlds/ufo50/barbuta/rules.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    regions["Mimic Room"].connect(regions["Boss Area"])
+
+    set_rule(world.get_location("Barbuta - Chest - G2"),
+             rule=lambda state: state.has(pin, player))
+    set_rule(world.get_location("Barbuta - Egg Shop - B6"),
+             rule=lambda state: has_money(100, state, player))
+    set_rule(world.get_location("Barbuta - Upper Shop Candy - C1"),
+             rule=lambda state: has_money(100, state, player))
+    set_rule(world.get_location("Barbuta - Upper Shop Umbrella - C1"),
+             rule=lambda state: has_money(100, state, player))  # todo: verify price
```
Comment (LeonarthCG) [inline worlds/ufo50/barbuta/rules.py]: Seems like it's only fifty bucks, wild.

**Review by LeonarthCG — COMMENTED** [review]
This is obviously progress over what I had.

There's a few details I'm not sure about, but otherwise, as you said, it'll just be a question of shaping this to whatever structure we come up with for games.

File: worlds/ufo50/barbuta/items.py (hunk 1 above)
Comment (ScipioWright) [reply worlds/ufo50/barbuta/items.py:24]: Prog + useful is meant to indicate "this is a really good progression item". The pin seems to unlock a lot of spots compared to other items, and the wand is something everyone probably wants

File: worlds/ufo50/barbuta/locations.py (hunk 2 above)
Comment (ScipioWright) [reply worlds/ufo50/barbuta/locations.py:30]: yeah I have no clue either, feel free to change it to whatever

File: worlds/ufo50/barbuta/items.py (hunk 3 above)
Comment (ScipioWright) [reply worlds/ufo50/barbuta/items.py]: Eh, I'll swap it to the internal name. Less confusing that way

**Pull request merged**